Sources/MongoSwift/MongoCollection+BulkWrite.swift

```diff
@@ -12,20 +12,23 @@ extension MongoCollection {
      * - Returns: a `BulkWriteResult`, or `nil` if the write concern is unacknowledged.
      *
      * - Throws:
-     *   - `MongoError.invalidArgument` if `requests` is empty
-     *   - `MongoError.bulkWriteError` if any error occurs while performing the writes
+     *   - `UserError.invalidArgumentError` if `requests` is empty.
+     *   - `ServerError.bulkWriteError` if any error occurs while performing the writes.
+     *   - `ServerError.commandError` if an error occurs that prevents the operation from being performed.
+     *   - `EncodingError` if an error occurs while encoding the `CollectionType` or the options to BSON.
      */
     @discardableResult
     public func bulkWrite(_ requests: [WriteModel], options: BulkWriteOptions? = nil) throws -> BulkWriteResult? {
         guard !requests.isEmpty else {
-            throw MongoError.invalidArgument(message: "requests cannot be empty")
+            throw UserError.invalidArgumentError(message: "requests cannot be empty")
         }
 
         let opts = try BSONEncoder().encode(options)
-        let bulk = BulkWriteOperation(collection: self._collection, opts: opts?.data)
+        let bulk = BulkWriteOperation(collection: self._collection, opts: opts)
 
         try requests.enumerated().forEach { index, model in
             try model.addToBulkWrite(bulk: bulk, index: index)
+            bulk.requests[index] = model
```

Comment:
Can you explain why this change was needed?

Reply:
Where does the requirement to access the requests in the error come from?

Reply:
The definition of `BulkWriteError` in the [crud spec](https://github.com/mongodb/specifications/blob/master/source/crud/crud.rst#error-handling) includes it.

Reply:
Well that would certainly justify it! I still think we should try to offload this onto the C driver, we shouldn't have to allocate these twice for potential error scenarios. As it stands, I'm paying the price for something I'll hopefully never see.

Reply:
IIRC, including the write model in the error is optional. The only requirement is that we report the index for individual write errors. I spoke to @patrickfreed about this last week as I saw he added this behavior but I didn't realize at the time that it would add quite a bit of overhead to the common no-error case.

I don't see how this can be offloaded to the C driver, as I don't believe it provides access to operations once they're added to a `mongoc_bulk_operation_t`. Might be best just to remove this altogether.

Reply:
The write model field in the error is an optional type (meaning it wouldn't be present in some cases, I guess. I'm not sure what those cases would be, though.), but it is _not_ marked as "NOT REQUIRED" as the fields which are optional for drivers to implement are. It seems, according to spec at least, the field is required. Whether or not it is worth it to us to conform to the spec in this case is another question, though.

Reply:
Python has a `dict` that represents a `BulkWriteErrror` which has an `'op'` field in it. It looks like Node has similar behavior.

Java, C#, Go, and Ruby don't report the request.

So none of the drivers report an actual `WriteModel`, but Node and Python do report some information about the request that caused the error.

Reply:
My thought about offloading onto the C driver was that we ask for an accessor for `mongoc_bulk_operation_t` operations. Outside of that, I'd argue that we amend the spec to clarify that this is optional, and not implement it until someone asks for it.

Reply:
I agree with @mbroadst

Reply:
To summarize:
I'll remove the request storage from `BulkWriteError` and file a follow-up SPEC ticket regarding making it NOT REQUIRED. If in the future libmongoc provides some sort of accessor on `bulk_operation_t` for getting operations by index or otherwise reports them in the errors, we can revisit this.

Reply:
Filed [SPEC-1209](https://jira.mongodb.org/browse/SPEC-1209)